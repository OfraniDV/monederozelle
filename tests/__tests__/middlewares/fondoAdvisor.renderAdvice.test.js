'use strict';

const { renderAdvice } = require('../../../middlewares/fondoAdvisor');

function buildBaseResult(overrides = {}) {
  const baseConfig = {
    cushion: 180000,
    sellRate: 452,
    minSellUsd: 40,
    liquidityBanks: ['BANDEC', 'MITRANSFER', 'METRO', 'BPA'],
    sellFeePct: 0,
    fxMarginPct: 0,
    sellRoundToUsd: 1,
    minKeepUsd: 0,
    limitMonthlyDefaultCup: 120000,
    limitMonthlyBpaCup: 120000,
    extendableBanks: ['BPA'],
    allocationBankOrder: ['BANDEC', 'MITRANSFER', 'METRO', 'BPA'],
  };

  const base = {
    activosCup: 412779,
    deudasCup: -275000,
    netoCup: 137779,
    cushionTarget: 180000,
    needCup: 42221,
    disponibles: 137779,
    plan: {
      sellTarget: { usd: 94, cupIn: 42488 },
      sellNow: { usd: 80, cupIn: 36160 },
      remainingCup: 6061,
      remainingUsd: 14,
      sellNet: 452,
    },
    projection: { negativosPost: 0, colchonPost: 173939 },
    liquidityByBank: {
      BANDEC: 322491,
      MITRANSFER: 53628,
      METRO: 23295,
      BPA: 13365,
    },
    config: baseConfig,
    deudaAbs: 275000,
    urgency: '🟠 PRIORITARIO',
    monthlyLimits: { cards: [], totals: { totalRemaining: 0, blocked: 0, extendable: 0 } },
    distributionNow: { assignments: [], leftover: 0, totalAssigned: 0 },
    distributionTarget: { assignments: [], leftover: 6061, totalAssigned: 0 },
    buyRateCup: 400,
    buyRateSource: 'db',
    sellRateSource: 'env',
    usdInventory: 180,
  };

  const merged = {
    ...base,
    ...overrides,
  };

  merged.plan = {
    ...base.plan,
    ...(overrides.plan || {}),
  };

  merged.projection = {
    ...base.projection,
    ...(overrides.projection || {}),
  };

  merged.config = {
    ...baseConfig,
    ...(overrides.config || {}),
  };

  merged.liquidityByBank = {
    ...base.liquidityByBank,
    ...(overrides.liquidityByBank || {}),
  };

  const overridesMonthly = overrides.monthlyLimits || {};
  merged.monthlyLimits = {
    cards: Array.isArray(overridesMonthly.cards) ? overridesMonthly.cards : base.monthlyLimits.cards,
    totals: {
      ...base.monthlyLimits.totals,
      ...(overridesMonthly.totals || {}),
    },
  };

  merged.distributionNow = {
    ...base.distributionNow,
    ...(overrides.distributionNow || {}),
  };

  merged.distributionTarget = overrides.hasOwnProperty('distributionTarget')
    ? overrides.distributionTarget
    : base.distributionTarget;

  return merged;
}

beforeEach(() => {
  jest.spyOn(console, 'log').mockImplementation(() => {});
  jest.spyOn(console, 'error').mockImplementation(() => {});
});

afterEach(() => {
  jest.restoreAllMocks();
});

describe('renderAdvice output formatting', () => {
  test('liquidez rápida muestra equivalentes en USD cuando hay tasa de compra', () => {
    const result = buildBaseResult();
    const blocks = renderAdvice(result);
    const liquidityBlock = blocks.find((b) => b.startsWith('🏦 <b>Liquidez rápida disponible</b>'));
    expect(liquidityBlock).toBeDefined();
    expect(liquidityBlock).toContain('• BANDEC: 322,491 CUP (≈ 806.23 USD)');
    expect(liquidityBlock).toContain('• MITRANSFER: 53,628 CUP (≈ 134.07 USD)');
    expect(liquidityBlock).toContain('• METRO: 23,295 CUP (≈ 58.24 USD)');
    expect(liquidityBlock).toContain('• BPA: 13,365 CUP (≈ 33.41 USD)');
  });

  test('estado actual elimina sufijo de tasa de compra en equivalentes USD', () => {
    const result = buildBaseResult();
    const message = renderAdvice(result).join('\n\n');
    expect(message).toContain('📊 <b>Estado actual</b>');
    expect(message).not.toContain('📊 <b>Estado actual CUP</b>');
    expect(message).toContain('ℹ️ <b>Indicadores</b>');
    expect(message).not.toContain('@ compra');
  });

  test('se eliminan bloques de tasas, equivalencias y parámetros', () => {
    const result = buildBaseResult();
    const message = renderAdvice(result).join('\n\n');
    expect(message).not.toContain('💱 <b>Tasas de referencia</b>');
    expect(message).not.toContain('🔄 <b>Equivalencias de referencia</b>');
    expect(message).not.toContain('📝 <b>Parámetros</b>');
  });

  test('snapshot completo mantiene formato HTML de liquidez', () => {
    const result = buildBaseResult();
    const message = renderAdvice(result).join('\n\n');
    expect(message).toMatchSnapshot();
  });

  test('liquidez rápida se mantiene solo en CUP cuando no hay tasa válida', () => {
    const result = buildBaseResult({ buyRateCup: null, buyRateSource: 'none' });
    const liquidityBlock = renderAdvice(result).find((b) =>
      b.startsWith('🏦 <b>Liquidez rápida disponible</b>')
    );
    expect(liquidityBlock).toBeDefined();
    expect(liquidityBlock).toContain('• BANDEC: 322,491 CUP');
    expect(liquidityBlock).not.toContain('USD');
  });
});
