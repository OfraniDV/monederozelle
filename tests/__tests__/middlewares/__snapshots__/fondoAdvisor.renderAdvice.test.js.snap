// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`renderAdvice output formatting snapshot completo mantiene formato HTML de liquidez 1`] = `
"🧮 <b>Asesor de Fondo</b>

🔴 URGENTE

━━━━━━━━━━━━━━━━━━

📊 <b>Estado actual</b>
• Activos: 412,779 CUP (≈ 1,031.95 USD)
• Deudas: -275,000 CUP (≈ -687.50 USD)
• Colchón actual: 137,779 CUP (≈ 344.45 USD)
• Neto: 137,779 CUP (≈ 344.45 USD)

ℹ️ <b>Indicadores</b>
• Deuda/Activos: 66.6%
• Avance colchón: 77% ▓▓▓▓▓▓▓▓░░

🎯 <b>Objetivo</b>
• Colchón objetivo: 180,000 CUP (≈ 450.00 USD)
• Necesidad adicional: 42,221 CUP (≈ 105.55 USD)

💵 <b>Inventario USD/Zelle</b>
• Total: 180 USD
• Disponible ahora: 180 USD


💸 <b>Venta requerida (Zelle)</b>
👉 Objetivo: vender 94 USD a 452 ⇒ +42,488 CUP (≈ 106.22 USD)
👉 Vende ahora: 80 USD ⇒ +36,160 CUP (≈ 90.40 USD)
• Faltante tras venta: 6,061 CUP (≈ 15.15 USD)


────────────────────────────────

🚦 <b>Límite mensual por tarjeta</b>

ℹ️ Límite mensual: Estándar 120,000 CUP (≈ 300.00 USD) • BPA 120,000 CUP (≈ 300.00 USD) (ampliable)

<pre>—</pre>

📍 <b>Sugerencia de destino</b>

<pre>Venta AHORA: 36,160 CUP (≈ 90.40 USD)
  Sin capacidad disponible
Objetivo total: 42,488 CUP (≈ 106.22 USD)
  Sin capacidad disponible
  ⚠️ CUP sin destino: 6,061 CUP (≈ 15.15 USD)</pre>

────────────────────────────────

🧾 <b>Proyección post-venta</b>
• Negativos: 0 CUP (≈ 0.00 USD)
• Colchón proyectado: 173,939 CUP (≈ 434.85 USD) &lt; 180,000 CUP (≈ 450.00 USD)

🏦 <b>Liquidez rápida disponible</b>
• BANDEC: 322,491 CUP (≈ 806.23 USD)
• MITRANSFER: 53,628 CUP (≈ 134.07 USD)
• METRO: 23,295 CUP (≈ 58.24 USD)
• BPA: 13,365 CUP (≈ 33.41 USD)

📐 <b>Explicación</b>
• Fórmula: necesidad = |deudas| + colchón − activos = 275,000 + 180,000 − 412,779 = 42,221 CUP
• Objetivo USD = ceil(necesidad / sellNet) (+ márgenes y redondeo)
• sellNet = floor(SELL × (1 − fee)) ⇒ SELL: 452 (fuente ENV)  fee: 0.00%
• Conversión: USD ≈ CUP / 400.00 (fuente DB)"
`;
